### PR TITLE
Improve spring config verification

### PR DESCRIPTION
**Pull request by Latropos**
## Summary

Each time a spring animation is triggered we check if provided config is valid. When it isn't we skip animation and change the animated view to its final position after waiting the time passed as `duration` (which may be zero)

This change fixes the following bug:
* Warning when spring used with zero duration (which should be a valid usage)
* No warning for negative mass

Additional benefits are:
* Invalid config warning informs what exactly is wrong
* Types are both more accurate and less verobse

Behaviour when duration is zero:

https://github.com/software-mansion/react-native-reanimated/assets/56199675/c16caa2a-9df0-4ba0-9890-4fe81b5b090d

### CODE REVIEW

**Comment by Latropos** [comment]
@bartlomiejbloniarz Do we have two ways of skipping the spring animation now? Do they have exactly the same behaviour?

**Comment by bartlomiejbloniarz** [comment]
@Latropos there is a slight difference between those behaviours. When reduced motion is enabled, the `toValue` is reached immediately. But when you pass an invalid config like `{duration:1000, dampingRatio:0}` the `toValue` will be reached after a second. Is this the intended behaviour?

**Comment by Latropos** [comment]
@bartlomiejbloniarz It was deliberate decision to wait the `duration` time, however now I think I wasn't a good idea and immediate change would be better.

File: src/reanimated2/animation/spring.ts
Comment (tomekzaw) [inline src/reanimated2/animation/spring.ts:61]: It's a little bit counterintuitive, I would expect that `validateConfig` returns whether the config is valid. 
```suggestion
    config.skipAnimation = !isConfigValid(config);
```

File: src/reanimated2/animation/spring.ts
Comment (tomekzaw) [inline src/reanimated2/animation/spring.ts:64]: If the current value is 100, the target value is 200 and duration is 0, do we set the final value in such case?

File: src/reanimated2/animation/springUtils.ts
Comment (tomekzaw) [inline src/reanimated2/animation/springUtils.ts:64]:
```suggestion
  let isConfigInvalid = false;
```

File: src/reanimated2/animation/springUtils.ts
Comment (tomekzaw) [inline src/reanimated2/animation/springUtils.ts:75]: In React world, `prop` is usually used here.
```suggestion
  ).forEach((prop) => {
```

File: src/reanimated2/animation/springUtils.ts
Comment (tomekzaw) [inline src/reanimated2/animation/springUtils.ts:79]:
```suggestion
    const value = config[prop];
    if (value <= 0) {
      isConfigInvalid = true;
      errorLog += `, ${prop} must be grater than zero but got ${value}`;
    }
```

File: src/reanimated2/animation/springUtils.ts
Comment (tomekzaw) [inline src/reanimated2/animation/springUtils.ts:63]: We should remove comma from here and instead add it to concrete error messages below.
```suggestion
  let errorMessage = '[Reanimated] Invalid spring config';
```
Apart from this, there's no need to allocate this string each time we call `withSpring`, maybe we could optimize the memory usage of the happy path here?

File: src/reanimated2/animation/springUtils.ts
Comment (tomekzaw) [inline src/reanimated2/animation/springUtils.ts:78]:
```suggestion
      errorLog += `, ${property} must be grater than zero but got ${config[property]}`;
```

File: src/reanimated2/animation/springUtils.ts
Comment (tomekzaw) [inline src/reanimated2/animation/springUtils.ts:84]:
```suggestion
    errorLog += `, duration can't be negative, got ${config.duration}`;
```

File: src/reanimated2/animation/springUtils.ts
Comment (tomekzaw) [inline src/reanimated2/animation/springUtils.ts:88]:
```suggestion
  if (isConfigInvalid) {
    console.warn(errorMessage);
```

**Review by tomekzaw — CHANGES_REQUESTED** [review]

File: src/reanimated2/animation/spring.ts
Comment (Latropos) [reply src/reanimated2/animation/spring.ts:64]: Yes it is.
In this line we set `skipAnimation` flag, which is used in this clause:
```
      if (config.skipAnimation) {
        // If we use duration we want to wait the provided time before stopping
        if (config.useDuration) return false;
        else {
          animation.current = toValue;
          animation.lastTimestamp = 0;
          return true;
        }
      }
```
Therefore the animation without duration sets the value inside this if condition and animation with `duration = 0` sets the final value in next iteration, here:

```javascript

      if (config.useDuration && timeFromStart >= config.duration) {
        animation.current = toValue;
        // clear lastTimestamp to avoid using stale value by the next spring animation that starts after this one
        animation.lastTimestamp = 0;
        return true;
      }
```

Comment (Latropos) [reply src/reanimated2/animation/spring.ts:64]: @tomekzaw  Now I regret the decision to wait the duration time for invalid config and I think we should always update the value immediately, WDYT?

Comment (tomekzaw) [reply src/reanimated2/animation/spring.ts:64]: Sounds ok

Comment (Latropos) [reply src/reanimated2/animation/spring.ts:64]: @tomekzaw Is this a change for this PR or for a separate one?

File: src/reanimated2/animation/spring.ts
Comment (tomekzaw) [inline src/reanimated2/animation/spring.ts:84]: There's no curly braces after `if`, can we add them and also check why linter doesn't complain about it? We should have `curly` rule enabled. Also, what's the conclusion when it comes to the expected behavior when config is invalid? Why does it depend on whether we use duration or not?
```suggestion
        if (config.useDuration) {
          return false;
        }
```

File: src/reanimated2/animation/springUtils.ts
Comment (tomekzaw) [inline src/reanimated2/animation/springUtils.ts:85]: The name of this method `isConfigValid` suggests that it has no side-effects, however it can call `console.warn`. Can we think of a better name here?

**Review by tomekzaw — CHANGES_REQUESTED** [review]

File: src/reanimated2/animation/spring.ts
Comment (Latropos) [reply src/reanimated2/animation/spring.ts:84]: I've thought we have decided that changing this behaviour should be done in separate PR (to avoid big changes on a reviewed PR), but since this is not a big change I'll delete it. It is because of the previous behaviour (we used to wait duration time before updating the value).

Comment (Latropos) [reply src/reanimated2/animation/spring.ts:84]: The rule 'curly' is only enabled in plugin

File: src/reanimated2/animation/springUtils.ts
Comment (Latropos) [reply src/reanimated2/animation/springUtils.ts:85]: I think that `checkIfConfigIsValid` may be better here.

File: src/reanimated2/animation/spring.ts
Comment (Latropos) [reply src/reanimated2/animation/spring.ts:84]: We can enable it (we have vialotion of this rule only in one file)

Comment (tomekzaw) [reply src/reanimated2/animation/spring.ts]: Thanks for checking this, let's do it (enable it everywhere)

File: src/reanimated2/animation/springUtils.ts
Comment (tomekzaw) [reply src/reanimated2/animation/springUtils.ts:85]: Sounds good

**Review by tomekzaw — APPROVED** [review]
Looks good to me. Let's wait for @piaskowyk's approval before merging.

**Review by piaskowyk — APPROVED** [review]

**Pull request merged**
